README.md
# OctaneWidgetsJS
Auto complete select, standard select and custom input based on select2 with custom events and easy configuration, with dependencies and self-work flow.
Able to pass triggered children on parent value by just one array!

#### Simple. Easy. Fast.

Mainly. Attached example is just an example in order to show how this works. 
As an example start typing country: United States, states will be loaded when loader is off.
In order to make request longer to be able to show loader at all - external source of states used.
Select Countries and select States are connected, also input age changes on Input year select.

## Core explanation:
#### 1. Regular select widget without autocomplete, but pre-loaded data

initialize widget

```JS
function SelectWidget(selector){};
```

sets custom loader name (set in the constructor but can be overridden as well)

```JS
SelectWidget.prototype.setLoaderName = function(name){};
```

shows loader

```JS
SelectWidget.prototype.loaderOn = function(){};
```

hides loader

```JS
SelectWidget.prototype.loaderOff = function(){};
```

sets field as hidden then no loader is triggered

```JS
SelectWidget.prototype.makeHidden = function(){};
```

makes field enabled (turns off disable attribute)

```JS
SelectWidget.prototype.enableNode = function(){};
```

makes field disabled

```JS
SelectWidget.prototype.disableNode = function(){};
```

set styles (also for loader, used in the constructor but can be overridden)

```JS
SelectWidget.prototype.setStyles = function(bgColor, leftDistance, topDistance){};
```

sets flag “select on preload“

```JS
SelectWidget.prototype.selectOnPreLoaded = function(selectOnly){};
```

sets url used in ajax call on parent field

```JS
SelectWidget.prototype.setUrl = function(url){};
```

load values and makes an ajax call on url for parent_id

```JS
SelectWidget.prototype.loadValues = function(parent_id){};
```

preloads item preset value and loads item for list dependant on that value if needed (e.g. to be able to select it on list among other values)

```JS
SelectWidget.prototype.preloadItem = function(value){};
```

populates values on list from given response and if value was already preset and exists within response, selects it

```JS
SelectWidget.prototype.populateValuesFromResponse = function(response, previousValue){};
```

loads entity as object in order to use its values later 
(e.g. on other select, can load entity values for further use, 
like selecting some select, triggers load of other properties etc.)

```JS
SelectWidget.prototype.loadEntityData = function(id, onSuccessCallback){};
```

triggers select on node

```JS
SelectWidget.prototype.selectValuesFromResponse = function(value){};
```

sets child widgets to be triggered by name into the property (differ it from calling, this is something different)

```JS
SelectWidget.prototype.setWidgetsToTriggerByNames = function(arrayOfNames){};
```

sets custom callback function after triggering has been done

```JS
SelectWidget.prototype.setCallbackOnTrigger = function(callbackFunction){};
```

sets “on click“ event

```JS
SelectWidget.prototype.setOnClick = function(){};
```

triggers click/change

```JS
SelectWidget.prototype.triggered = function(event){};
```

triggers children set in “setWidgetsToTriggerByNames()“

```JS
SelectWidget.prototype.triggerChildren = function(parentValue){};
```

sets flag “isDisabledWhenLoaded“ on true or false (false by default)

```JS
SelectWidget.prototype.disableWhenLoaded = function(isDisabledWhenLoaded){};
```

used when ajax call returns with success

```JS
SelectWidget.prototype.runSuccessfulResponse = function(response){};
```

used when ajax call returns with an error or not expected result

```JS
SelectWidget.prototype.runErrorResponse = function(exception){};
```


#### 2. Autocomplete select widget without pre-loaded data

initialize widget

```JS
function SelectAutoCompleteWidget(selector){};
```

sets custom loader name (set in the constructor but can be overridden as well)

```JS
SelectAutoCompleteWidget.prototype.setLoaderName = function(name){};
```

shows loader

```JS
SelectAutoCompleteWidget.prototype.loaderOn = function(){};
```

hides loader

```JS
SelectAutoCompleteWidget.prototype.loaderOff = function(){};
```

sets field as hidden then no loader is triggered

```JS
SelectAutoCompleteWidget.prototype.makeHidden = function(){};
```

makes field enabled (turns off disable attribute)

```JS
SelectAutoCompleteWidget.prototype.enableNode = function(){};
```

makes field disabled

```JS
SelectAutoCompleteWidget.prototype.disableNode = function(){};
```

set styles (also for loader, used in the constructor but can be overridden)

```JS
SelectAutoCompleteWidget.prototype.setStyles = function(bgColor, leftDistance, topDistance){};
```

sets flag “select on preload“

```JS
SelectAutoCompleteWidget.prototype.selectOnPreLoaded = function(selectOnly){};
```

sets url used in ajax call on parent field

```JS
SelectAutoCompleteWidget.prototype.setUrl = function(url){};
```

load values and makes an ajax call on url for parent_id

```JS
SelectAutoCompleteWidget.prototype.loadValues = function(parent_id){};
```

preloads item preset value and loads item for list dependant on that value if needed (e.g. to be able to select it on list among other values)

```JS
SelectAutoCompleteWidget.prototype.preloadItem = function(value){};
```

sets autocomplete process results function in case of change for given url

```JS
SelectAutoCompleteWidget.prototype.prepareAutoCompleteListener = function(){};
```

not implemented as method above is enough

```JS
SelectAutoCompleteWidget.prototype.processAutoCompleteValues = function(data){};
```

populates values on list from given response and if value was already preset and exists within response, selects it

```JS
SelectAutoCompleteWidget.prototype.populateValuesFromResponse = function(response){};
```

loads entity as object in order to use its values later 

```JS
SelectAutoCompleteWidget.prototype.loadEntityData = function(id, onSuccessCallback){};
```

values for further use, like selecting someWidget, 
triggers load of other properties etc.)

```JS
SelectAutoCompleteWidget.prototype.selectValuesFromResponse = function(value){};
```

sets child widgets to be triggered by name into the property (differ it from calling, this is something different)

```JS
SelectAutoCompleteWidget.prototype.setWidgetsToTriggerByNames = function(arrayOfNames){};
```

sets custom callback function after triggering has been done

```JS
SelectAutoCompleteWidget.prototype.setCallbackOnTrigger = function(callbackFunction){};
```

sets “on click“ event

```JS
SelectAutoCompleteWidget.prototype.setOnClick = function(){};
```

triggers click/change

```JS
SelectAutoCompleteWidget.prototype.triggered = function(event){};
```

triggers children set in “setWidgetsToTriggerByNames()“

```JS
SelectAutoCompleteWidget.prototype.triggerChildren = function(parentValue){};
```

sets flag “isDisabledWhenLoaded“ on true or false (false by default)

```JS
SelectAutoCompleteWidget.prototype.disableWhenLoaded = function(isDisabledWhenLoaded){};
```

used when ajax call returns with success

```JS
SelectAutoCompleteWidget.prototype.runSuccessfulResponse = function(response){};
```

used when ajax call returns with an error or not expected result

```JS
SelectAutoCompleteWidget.prototype.runErrorResponse = function(exception, type){};
```


#### 3. Input widget

initialize widget

```JS
function InputWidget(selector){};
```

sets custom loader name (set in the constructor but can be overridden as well)

```JS
InputWidget.prototype.setLoaderName = function(name){};
```

shows loader

```JS
InputWidget.prototype.loaderOn = function(){};
```

hides loader

```JS
InputWidget.prototype.loaderOff = function(){};
```

sets field as hidden then no loader is triggered

```JS
InputWidget.prototype.makeHidden = function(){};
```

makes field enabled (turns off disable attribute)

```JS
InputWidget.prototype.enableNode = function(){};
```

makes field disabled

```JS
InputWidget.prototype.disableNode = function(){};
```

set styles (also for loader, used in the constructor but can be overridden)

```JS
InputWidget.prototype.setStyles = function(bgColor, leftDistance, topDistance){};
```

sets url used in ajax call on parent field

```JS
InputWidget.prototype.setUrl = function(url){};
```

populate values given from response

```JS
InputWidget.prototype.populateValuesFromResponse = function(responseValue){};
```

preloads item on list

```JS
InputWidget.prototype.preloadItem = function(value){};
```

triggers field (changes its value) and it’s children depending on loaded value

```JS
InputWidget.prototype.triggered = function(event){};
```

triggers children by parent’s value

```JS
InputWidget.prototype.triggerChildren = function(parentValue){};
```

sets flag “isDisabledWhenLoaded“ on true or false (false by default)

```JS
InputWidget.prototype.disableWhenLoaded = function(isDisabledWhenLoaded){};
```

sets childer fields to be triggered on parent change by their name

```JS
InputWidget.prototype.setWidgetsToTriggerByNames = function(arrayOfNames){};
```

sets callback function after parent field triggered

```JS
InputWidget.prototype.setCallbackOnTrigger = function(callbackFunction){};
```

used when ajax call returns with success

```JS
InputWidget.prototype.runSuccessfulResponse = function(response){};
```

used when ajax call returns with an error or not expected result

```JS
InputWidget.prototype.runErrorResponse = function(exception){};
```


## Usage examples:
  Example flow:

  First you need to define global settings like main variable and urls:
  
```JS
  const GLOBAL_JS = {};
  const URLS = {
    'getExamplestuffOne' : "{{URL::route('urlToGetExamplestuffOne')}}",
    'getExamplestuffTwo' : "{{URL::route('getExamplestuffTwo')}}",
  };
```

  Property `GLOBAL_JS` will be used to store widgets instances for global access within view from any place, URLS will be storing urls in order not to set them thousand times in code.
  
  Everything is happening within:
  
```JS
  $(document).ready( function () {
    //...
  });
```

  Widget’s fields are based on ```JS
Select2`, so on ready it has to be initialized on every related field:

```JS
  $('.select2').select2({
    placeholder : $(this).data('placeholder'),
  });
```

  Since there is no general loader in css, loader for each field is defined “in fly” (yes, that needs improvement but there was no time to do so), so the general color of loader has to be set:

```JS
    GLOBAL_JS.LOADER_COLOR = '#139c9b'; // general loader color
    GLOBAL_JS.INP_TOP      = '15px';    // general top spacing
    GLOBAL_JS.INP_LFT      = '30%';     // general left spacing
    GLOBAL_JS.AJAX_FAILED  = false;     // if exception thrown, this is true
```
SelectAutoCompleteWidget: AutoComplete select on some_select_id field example:

```JS
  GLOBAL_JS.someWidget = new SelectAutoCompleteWidget('select[name="some_select_id"]');
  GLOBAL_JS.someWidget
    .setUrl(URLS.getSomeStuff)
    .disableWhenLoaded(GLOBAL_JS.I_WANT_THIS_TO_BE_DISABLED_WHEN_LOADED)
    .prepareAutoCompleteListener();
```
  Triggers entity data to be loaded on change (if children to trigger set / look below).
  Load full entity data on select change and get values from callback after success.
You can trigger either inputs that has been defined as widget  (someOtherWidget) or simple form inputs (not-widget-input)
```JS
  GLOBAL_JS.someWidget.setCallbackOnTrigger(function(){
      let INSTANCE = this;
    
      INSTANCE.loadEntityData(INSTANCE.widgetNode.val(), function(loadedEntityData){
      INSTANCE.widgetSelector.currentEntity = loadedEntityData;
      
      if(typeof loadedEntityData.id !== 'undefined'){
        GLOBAL_JS.someOtherWidget.widgetNode.val(loadedEntityData.some_value).trigger('change');
        GLOBAL_JS.someOtherWidget.triggered();
      }
  
      if(typeof loadedEntityData.some_value !== 'undefined')
        $('input[name=not-widget-input]').val(loadedEntityData.some_value).trigger('change');
  
    });
  });
```

CHANGELOG.md
##CHANGELOG


####v1.0
Initial version

ajax.php
<?php

include 'countries.php';
//include 'us_states.php';

/** Example ajax responses */
$result = [];
if(isset($_GET['t'])){

    switch ($_GET['t']){
        case "c":
            if(isset($_GET['phrase'])){
                $phrase = $_GET['phrase'];
                $keysFromArray = array_keys(array_filter($countries, function($var) use ($phrase){
                    return strpos(strtolower($var), strtolower($phrase)) !== false;
                }));
                $result = array_intersect_key($countries, array_flip($keysFromArray));
            }
            break;
        case "s":
            if(isset($_GET['id']) && strtolower($_GET['id']) === 'us'){
                //$result = $usStates;
                $result = getFromUrl('https://gist.githubusercontent.com/mshafrir/2646763/raw/8b0dbb93521f5d6889502305335104218454c2bf/states_hash.json');
                $result = json_decode($result);
            }else{
                $result = ["NA" => "Unknown"];
            }

            break;
        default: break;

    }
}

$keyValue = [];
foreach($result as $key => $value){
    $keyValue[] = ['key' => $key, 'value' => $value];
}

echo json_encode($keyValue);
die();

function getFromUrl($url)
{
    if(empty($url)) return '';
    $cURLConnection = curl_init();

    curl_setopt($cURLConnection, CURLOPT_URL, $url);
    curl_setopt($cURLConnection, CURLOPT_RETURNTRANSFER, true);

    $result = curl_exec($cURLConnection);
    curl_close($cURLConnection);

    return $result;
}


index.php

<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description" content="">
    <meta name="author" content="">
    <!--<link rel="icon" href="/docs/4.0/assets/img/favicons/favicon.ico">-->

    <title>OctaneWidgetJS example form</title>

    <link rel="canonical" href="https://getbootstrap.com/docs/4.0/examples/checkout/">

    <!-- Bootstrap core CSS -->
    <link href="css/bootstrap.min.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="css/select2.min.css" rel="stylesheet"/>
    <link href="css/octane.css" rel="stylesheet"/>
</head>

<body class="bg-light">

<div class="container">
    <div class="py-5 text-center">
        <h2>Example form</h2>
    </div>

    <div class="row">
        <div class="col-md-6 offset-md-3">
            <h4 class="mb-3">Address</h4>
            <form class="needs-validation" novalidate>
                <div class="row">
                    <div class="col-md-6 mb-3">
                        <label for="firstName">First name</label>
                        <input type="text" class="form-control" id="firstName" placeholder="" value="" required>
                        <div class="invalid-feedback">
                            Valid first name is required.
                        </div>
                    </div>
                    <div class="col-md-6 mb-3">
                        <label for="lastName">Last name</label>
                        <input type="text" class="form-control" id="lastName" placeholder="" value="" required>
                        <div class="invalid-feedback">
                            Valid last name is required.
                        </div>
                    </div>
                </div>

                <div class="mb-3">
                    <label for="email">Email <span class="text-muted">(Optional)</span></label>
                    <input type="email" class="form-control" id="email" placeholder="you@example.com">
                    <div class="invalid-feedback">
                        Please enter a valid email address for shipping updates.
                    </div>
                </div>

                <div class="mb-3">
                    <label for="address">Address</label>
                    <input type="text" class="form-control" id="address" placeholder="Some street 13B" required>
                    <div class="invalid-feedback">
                        Please enter your shipping address.
                    </div>
                </div>

                <div class="row">
                    <div class="col-md-5 mb-3">
                        <label for="country">Country</label>
                        <select class="select2 custom-select d-block w-100" id="country" name="country" required>
                            <option value="">Choose...</option>
                        </select>
                        <div class="invalid-feedback">
                            Please select a valid country.
                        </div>
                    </div>
                    <div class="col-md-4 mb-3">
                        <label for="state">State / Province</label>
                        <select class="select2 custom-select d-block w-100" id="state" name="state" required>
                            <option value="">Choose...</option>
                        </select>
                        <div class="invalid-feedback">
                            Please provide a valid state.
                        </div>
                    </div>
                    <div class="col-md-3 mb-3">
                        <label for="zip">Zip</label>
                        <input type="text" class="form-control" id="zip" name="zip" placeholder="" required>
                        <div class="invalid-feedback">
                            Zip code required.
                        </div>
                    </div>
                </div>

                <h4 class="mb-3">D.O.B.</h4>
                <div class="row">
                    <div class="col-md-5 mb-3">
                        <label for="year">Year</label>
                        <select class="select2 custom-select d-block w-100" id="year" name="year" required>
                            <option value="">Choose...</option>
                            <?php for($i = date('Y'); $i>1900; $i--): ?>
                                <option value="<?= $i ?>"><?= $i ?></option>
                            <?php endfor; ?>
                        </select>
                        <div class="invalid-feedback">
                            Please select a valid year.
                        </div>
                    </div>

                    <div class="col-md-4 mb-3">
                        <label for="month">Month</label>
                        <select class="select2 custom-select d-block w-100" id="month" name="month" required>
                            <option value="">Choose...</option>
                            <?php for($m = 1; $m <= 12; ++$m): ?>
                                <option value="<?= $m  ?>"><?= date('F', mktime(0, 0, 0, $m, 1)) ?></option>
                            <?php endfor; ?>
                        </select>
                        <div class="invalid-feedback">
                            Please provide a valid month.
                        </div>
                    </div>
                    <div class="col-md-3 mb-3">
                        <label for="mondayth">Day</label>
                        <select class="select2 custom-select d-block w-100" id="day" name="day" required>
                            <option value="">Choose...</option>
                            <?php for($d = 1; $d <= 31; ++$d): ?>
                                <option value="<?= $d  ?>"><?= $d ?></option>
                            <?php endfor; ?>
                        </select>
                        <div class="invalid-feedback">
                            Please provide a valid day.
                        </div>
                    </div>
                </div>

                <div class="row">
                    <div class="col-md-3 mb-3">
                        <label for="age">Age</label>
                        <input type="text" class="form-control" id="age" name="age" placeholder="0" disabled>
                    </div>
                </div>
                <hr class="mb-4">
                <button class="btn btn-primary btn-lg btn-block" type="submit">Continue to checkout</button>
            </form>
        </div>
    </div>

    <footer class="my-5 pt-5 text-muted text-center text-small">
        <p class="mb-1">&copy; 2020 Company Name</p>
        <ul class="list-inline">
            <li class="list-inline-item"><a href="#">Privacy</a></li>
            <li class="list-inline-item"><a href="#">Terms</a></li>
            <li class="list-inline-item"><a href="#">Support</a></li>
        </ul>
    </footer>
</div>

<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<!--<script src="https://code.jquery.com/jquery-3.1.1.min.js" crossorigin="anonymous"></script>-->
<script src="js/jquery-3.1.1.min.js"></script>
<script>window.jQuery || document.write('<script src="js/jquery-slim.min.js"><\/script>')</script>
<script src="js/popper.min.js"></script>
<script src="js/bootstrap.min.js"></script>
<script src="js/holder.min.js"></script>
<script src="js/select2.min.js"></script>

<script src="js/octane/inputWidget.js"></script>
<script src="js/octane/selectWidget.js"></script>
<script src="js/octane/selectAutoCompleteWidget.js"></script>

<script src="js/core.js"></script>

<script type="text/javascript">

</script>
</body>
</html>


countries.php
<?php
$countries =
    [
        "AF" => "Afghanistan",
        "AL" => "Albania",
        "DZ" => "Algeria",
        "AS" => "American Samoa",
        "AD" => "Andorra",
        "AO" => "Angola",
        "AI" => "Anguilla",
        "AQ" => "Antarctica",
        "AG" => "Antigua and Barbuda",
        "AR" => "Argentina",
        "AM" => "Armenia",
        "AW" => "Aruba",
        "AU" => "Australia",
        "AT" => "Austria",
        "AZ" => "Azerbaijan",
        "BS" => "Bahamas",
        "BH" => "Bahrain",
        "BD" => "Bangladesh",
        "BB" => "Barbados",
        "BY" => "Belarus",
        "BE" => "Belgium",
        "BZ" => "Belize",
        "BJ" => "Benin",
        "BM" => "Bermuda",
        "BT" => "Bhutan",
        "BO" => "Bolivia",
        "BA" => "Bosnia and Herzegovina",
        "BW" => "Botswana",
        "BV" => "Bouvet Island",
        "BR" => "Brazil",
        "IO" => "British Indian Ocean Territory",
        "BN" => "Brunei Darussalam",
        "BG" => "Bulgaria",
        "BF" => "Burkina Faso",
        "BI" => "Burundi",
        "KH" => "Cambodia",
        "CM" => "Cameroon",
        "CA" => "Canada",
        "CV" => "Cape Verde",
        "KY" => "Cayman Islands",
        "CF" => "Central African Republic",
        "TD" => "Chad",
        "CL" => "Chile",
        "CN" => "China",
        "CX" => "Christmas Island",
        "CC" => "Cocos (Keeling) Islands",
        "CO" => "Colombia",
        "KM" => "Comoros",
        "CG" => "Congo",
        "CD" => "Congo, the Democratic Republic of the",
        "CK" => "Cook Islands",
        "CR" => "Costa Rica",
        "CI" => "Cote D'Ivoire",
        "HR" => "Croatia",
        "CU" => "Cuba",
        "CY" => "Cyprus",
        "CZ" => "Czech Republic",
        "DK" => "Denmark",
        "DJ" => "Djibouti",
        "DM" => "Dominica",
        "DO" => "Dominican Republic",
        "EC" => "Ecuador",
        "EG" => "Egypt",
        "SV" => "El Salvador",
        "GQ" => "Equatorial Guinea",
        "ER" => "Eritrea",
        "EE" => "Estonia",
        "ET" => "Ethiopia",
        "FK" => "Falkland Islands (Malvinas)",
        "FO" => "Faroe Islands",
        "FJ" => "Fiji",
        "FI" => "Finland",
        "FR" => "France",
        "GF" => "French Guiana",
        "PF" => "French Polynesia",
        "TF" => "French Southern Territories",
        "GA" => "Gabon",
        "GM" => "Gambia",
        "GE" => "Georgia",
        "DE" => "Germany",
        "GH" => "Ghana",
        "GI" => "Gibraltar",
        "GR" => "Greece",
        "GL" => "Greenland",
        "GD" => "Grenada",
        "GP" => "Guadeloupe",
        "GU" => "Guam",
        "GT" => "Guatemala",
        "GN" => "Guinea",
        "GW" => "Guinea-Bissau",
        "GY" => "Guyana",
        "HT" => "Haiti",
        "HM" => "Heard Island and Mcdonald Islands",
        "VA" => "Holy See (Vatican City State)",
        "HN" => "Honduras",
        "HK" => "Hong Kong",
        "HU" => "Hungary",
        "IS" => "Iceland",
        "IN" => "India",
        "ID" => "Indonesia",
        "IR" => "Iran, Islamic Republic of",
        "IQ" => "Iraq",
        "IE" => "Ireland",
        "IL" => "Israel",
        "IT" => "Italy",
        "JM" => "Jamaica",
        "JP" => "Japan",
        "JO" => "Jordan",
        "KZ" => "Kazakhstan",
        "KE" => "Kenya",
        "KI" => "Kiribati",
        "KP" => "Korea, Democratic People's Republic of",
        "KR" => "Korea, Republic of",
        "KW" => "Kuwait",
        "KG" => "Kyrgyzstan",
        "LA" => "Lao People's Democratic Republic",
        "LV" => "Latvia",
        "LB" => "Lebanon",
        "LS" => "Lesotho",
        "LR" => "Liberia",
        "LY" => "Libyan Arab Jamahiriya",
        "LI" => "Liechtenstein",
        "LT" => "Lithuania",
        "LU" => "Luxembourg",
        "MO" => "Macao",
        "MK" => "Macedonia, the Former Yugoslav Republic of",
        "MG" => "Madagascar",
        "MW" => "Malawi",
        "MY" => "Malaysia",
        "MV" => "Maldives",
        "ML" => "Mali",
        "MT" => "Malta",
        "MH" => "Marshall Islands",
        "MQ" => "Martinique",
        "MR" => "Mauritania",
        "MU" => "Mauritius",
        "YT" => "Mayotte",
        "MX" => "Mexico",
        "FM" => "Micronesia, Federated States of",
        "MD" => "Moldova, Republic of",
        "MC" => "Monaco",
        "MN" => "Mongolia",
        "MS" => "Montserrat",
        "MA" => "Morocco",
        "MZ" => "Mozambique",
        "MM" => "Myanmar",
        "NA" => "Namibia",
        "NR" => "Nauru",
        "NP" => "Nepal",
        "NL" => "Netherlands",
        "AN" => "Netherlands Antilles",
        "NC" => "New Caledonia",
        "NZ" => "New Zealand",
        "NI" => "Nicaragua",
        "NE" => "Niger",
        "NG" => "Nigeria",
        "NU" => "Niue",
        "NF" => "Norfolk Island",
        "MP" => "Northern Mariana Islands",
        "NO" => "Norway",
        "OM" => "Oman",
        "PK" => "Pakistan",
        "PW" => "Palau",
        "PS" => "Palestinian Territory, Occupied",
        "PA" => "Panama",
        "PG" => "Papua New Guinea",
        "PY" => "Paraguay",
        "PE" => "Peru",
        "PH" => "Philippines",
        "PN" => "Pitcairn",
        "PL" => "Poland",
        "PT" => "Portugal",
        "PR" => "Puerto Rico",
        "QA" => "Qatar",
        "RE" => "Reunion",
        "RO" => "Romania",
        "RU" => "Russian Federation",
        "RW" => "Rwanda",
        "SH" => "Saint Helena",
        "KN" => "Saint Kitts and Nevis",
        "LC" => "Saint Lucia",
        "PM" => "Saint Pierre and Miquelon",
        "VC" => "Saint Vincent and the Grenadines",
        "WS" => "Samoa",
        "SM" => "San Marino",
        "ST" => "Sao Tome and Principe",
        "SA" => "Saudi Arabia",
        "SN" => "Senegal",
        "CS" => "Serbia and Montenegro",
        "SC" => "Seychelles",
        "SL" => "Sierra Leone",
        "SG" => "Singapore",
        "SK" => "Slovakia",
        "SI" => "Slovenia",
        "SB" => "Solomon Islands",
        "SO" => "Somalia",
        "ZA" => "South Africa",
        "GS" => "South Georgia and the South Sandwich Islands",
        "ES" => "Spain",
        "LK" => "Sri Lanka",
        "SD" => "Sudan",
        "SR" => "Suriname",
        "SJ" => "Svalbard and Jan Mayen",
        "SZ" => "Swaziland",
        "SE" => "Sweden",
        "CH" => "Switzerland",
        "SY" => "Syrian Arab Republic",
        "TW" => "Taiwan, Province of China",
        "TJ" => "Tajikistan",
        "TZ" => "Tanzania, United Republic of",
        "TH" => "Thailand",
        "TL" => "Timor-Leste",
        "TG" => "Togo",
        "TK" => "Tokelau",
        "TO" => "Tonga",
        "TT" => "Trinidad and Tobago",
        "TN" => "Tunisia",
        "TR" => "Turkey",
        "TM" => "Turkmenistan",
        "TC" => "Turks and Caicos Islands",
        "TV" => "Tuvalu",
        "UG" => "Uganda",
        "UA" => "Ukraine",
        "AE" => "United Arab Emirates",
        "GB" => "United Kingdom",
        "US" => "United States",
        "UM" => "United States Minor Outlying Islands",
        "UY" => "Uruguay",
        "UZ" => "Uzbekistan",
        "VU" => "Vanuatu",
        "VE" => "Venezuela",
        "VN" => "Viet Nam",
        "VG" => "Virgin Islands, British",
        "VI" => "Virgin Islands, U.s.",
        "WF" => "Wallis and Futuna",
        "EH" => "Western Sahara",
        "YE" => "Yemen",
        "ZM" => "Zambia",
        "ZW" => "Zimbabwe"
];
?>

us_states.php
<?php

$usStates = [
    'AL'=>"Alabama",
    'AK'=>"Alaska",
    'AZ'=>"Arizona",
    'AR'=>"Arkansas",
    'CA'=>"California",
    'CO'=>"Colorado",
    'CT'=>"Connecticut",
    'DE'=>"Delaware",
    'DC'=>"District Of Columbia",
    'FL'=>"Florida",
    'GA'=>"Georgia",
    'HI'=>"Hawaii",
    'ID'=>"Idaho",
    'IL'=>"Illinois",
    'IN'=>"Indiana",
    'IA'=>"Iowa",
    'KS'=>"Kansas",
    'KY'=>"Kentucky",
    'LA'=>"Louisiana",
    'ME'=>"Maine",
    'MD'=>"Maryland",
    'MA'=>"Massachusetts",
    'MI'=>"Michigan",
    'MN'=>"Minnesota",
    'MS'=>"Mississippi",
    'MO'=>"Missouri",
    'MT'=>"Montana",
    'NE'=>"Nebraska",
    'NV'=>"Nevada",
    'NH'=>"New Hampshire",
    'NJ'=>"New Jersey",
    'NM'=>"New Mexico",
    'NY'=>"New York",
    'NC'=>"North Carolina",
    'ND'=>"North Dakota",
    'OH'=>"Ohio",
    'OK'=>"Oklahoma",
    'OR'=>"Oregon",
    'PA'=>"Pennsylvania",
    'RI'=>"Rhode Island",
    'SC'=>"South Carolina",
    'SD'=>"South Dakota",
    'TN'=>"Tennessee",
    'TX'=>"Texas",
    'UT'=>"Utah",
    'VT'=>"Vermont",
    'VA'=>"Virginia",
    'WA'=>"Washington",
    'WV'=>"West Virginia",
    'WI'=>"Wisconsin",
    'WY'=>"Wyoming"
];
?>

js/core.js

const URLS = {
    'getCountries'  : window.location.href + "/ajax.php?t=c",
    'getStates'     : window.location.href + "/ajax.php?t=s",
};

const TRANS = {
    429 : "Too many requests",
};

const GLOBAL_JS = {};

$(document).ready( function () {

    let $selectTwo = $('.select2');
    $selectTwo.select2(
        {
            placeholder : $(this).data('placeholder'),
            theme: "default custom-select", //select2 select2-container select2-container--default
        }
    ).data('select2').$dropdown.addClass('custom-select');

    $selectTwo.on('change', function(e) {
        $(e.currentTarget).parent().find(".select2-selection__rendered").attr('title', "Choose");
    });

    // $('.select2').on("change", function (e) {
    //     $(e.target).siblings('.select2-container').addClass('custom-select');
    //     console.log($(e.target).siblings('.select2-container'));
    // });

    GLOBAL_JS.LOADER_COLOR = '#139c9b';
    GLOBAL_JS.INP_TOP      = '15px';
    GLOBAL_JS.INP_LFT      = '30%';
    GLOBAL_JS.AJAX_FAILED  = false;

    // 1. country will load states
    GLOBAL_JS.countryWidget = new SelectAutoCompleteWidget('select[name="country"]');
    GLOBAL_JS.countryWidget
        .setUrl(URLS.getCountries)
        .setOldValue("0")
        .prepareAutoCompleteListener();

    // 2. state by country
    GLOBAL_JS.stateWidget = new SelectWidget('select[name="state"]');
    GLOBAL_JS.stateWidget
        .setUrl(URLS.getStates);
    //.prepareAutoCompleteListener();

    // 3. year will calculate age
    GLOBAL_JS.yearWidget = new SelectWidget('select[name="year"]');

    // 4. months will load days
    GLOBAL_JS.monthWidget = new SelectWidget('select[name="month"]');

    // 5. days depends on month
    GLOBAL_JS.dayWidget = new SelectWidget('select[name="day"]');

    // 6. age input
    GLOBAL_JS.ageWidget = new InputWidget('input[name="age"]');
    GLOBAL_JS.ageWidget
        .disableNode();


    // OTHER
    // Prepare triggers for children
    GLOBAL_JS.countryWidget.setWidgetsToTriggerByNames(['state']);
    //GLOBAL_JS.yearWidget.setWidgetsToTriggerByNames(['age']);

    // Prepare event listeners (needed even if has any callback below)
    GLOBAL_JS.countryWidget.setOnClick();
    GLOBAL_JS.yearWidget.setOnClick();
    GLOBAL_JS.monthWidget.setOnClick();

    // Triggers with callback
    GLOBAL_JS.countryWidget.setCallbackOnTrigger(function(){
        let INSTANCE = this;
        console.log("some custom callback after being triggered");
    });

    GLOBAL_JS.yearWidget.setCallbackOnTrigger(function(){
        let INSTANCE = this;
        if(typeof INSTANCE.widgetNode !== 'undefined' && INSTANCE.widgetNode.val() > 0){
            let chosenYear = INSTANCE.widgetNode.val();
            let currentDate = new Date();

            GLOBAL_JS.ageWidget.loaderOn();
            GLOBAL_JS.ageWidget.widgetNode.val(parseInt(currentDate.getFullYear()) - parseInt(chosenYear));
            GLOBAL_JS.ageWidget.loaderOff();
        }
    });

    // THIS IS JUST AN EXAMPLE, Feb length is general
    GLOBAL_JS.monthWidget.setCallbackOnTrigger(function(){
        let INSTANCE = this;
        if(typeof INSTANCE.widgetNode !== 'undefined' && INSTANCE.widgetNode.val() > 0 && typeof GLOBAL_JS.dayWidget !== 'undefined'){
            GLOBAL_JS.dayWidget.loaderOn();

            let chosenMonth = parseInt(INSTANCE.widgetNode.val());
            let dayOptions  = GLOBAL_JS.dayWidget.widgetNode.find('option').remove();

            let monthsWithThirtyOne = [1,3,5, 7, 8, 10, 12];
            let options = '';

            if(chosenMonth === 2){
                options = generateDayOptions(28);
            }else if(monthsWithThirtyOne.indexOf(chosenMonth) > -1){
                options = generateDayOptions(31);
            }else{
                options = generateDayOptions(30);
            }

            GLOBAL_JS.dayWidget.widgetNode.html(options);
            GLOBAL_JS.dayWidget.loaderOff();

        }
    });

    function generateDayOptions(upToDayIncluding) {
        let html = '';
        for (let i = 0; i < upToDayIncluding; i++) {
            let day = i + 1;

            if(day <= upToDayIncluding){
                html += '<option value="' + day + '">' + day + "</option>";
            }
        }

        return html;
    }

});